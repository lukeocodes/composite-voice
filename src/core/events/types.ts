/**
 * Event type definitions for the CompositeVoice SDK
 */

import type { AudioChunk, AudioMetadata } from '../types/audio';

/**
 * Agent state
 */
export type AgentState = 'idle' | 'ready' | 'listening' | 'thinking' | 'speaking' | 'error';

/**
 * Base event interface
 */
export interface BaseEvent {
  /** Timestamp when the event occurred */
  timestamp: number;
  /** Optional event metadata */
  metadata?: Record<string, unknown>;
}

/**
 * Transcription events
 */
export interface TranscriptionStartEvent extends BaseEvent {
  type: 'transcription.start';
}

export interface TranscriptionInterimEvent extends BaseEvent {
  type: 'transcription.interim';
  text: string;
  confidence?: number;
}

export interface TranscriptionFinalEvent extends BaseEvent {
  type: 'transcription.final';
  text: string;
  confidence?: number;
}

export interface TranscriptionErrorEvent extends BaseEvent {
  type: 'transcription.error';
  error: Error;
  recoverable: boolean;
}

export type TranscriptionEvent =
  | TranscriptionStartEvent
  | TranscriptionInterimEvent
  | TranscriptionFinalEvent
  | TranscriptionErrorEvent;

/**
 * LLM events
 */
export interface LLMStartEvent extends BaseEvent {
  type: 'llm.start';
  prompt: string;
}

export interface LLMChunkEvent extends BaseEvent {
  type: 'llm.chunk';
  chunk: string;
  accumulated: string;
}

export interface LLMCompleteEvent extends BaseEvent {
  type: 'llm.complete';
  text: string;
  tokensUsed?: number;
}

export interface LLMErrorEvent extends BaseEvent {
  type: 'llm.error';
  error: Error;
  recoverable: boolean;
}

export type LLMEvent = LLMStartEvent | LLMChunkEvent | LLMCompleteEvent | LLMErrorEvent;

/**
 * TTS events
 */
export interface TTSStartEvent extends BaseEvent {
  type: 'tts.start';
  text: string;
}

export interface TTSAudioEvent extends BaseEvent {
  type: 'tts.audio';
  chunk: AudioChunk;
}

export interface TTSMetadataEvent {
  type: 'tts.metadata';
  timestamp: number;
  metadata: AudioMetadata;
}

export interface TTSCompleteEvent extends BaseEvent {
  type: 'tts.complete';
}

export interface TTSErrorEvent extends BaseEvent {
  type: 'tts.error';
  error: Error;
  recoverable: boolean;
}

export type TTSEvent =
  | TTSStartEvent
  | TTSAudioEvent
  | TTSMetadataEvent
  | TTSCompleteEvent
  | TTSErrorEvent;

/**
 * Agent lifecycle events
 */
export interface AgentReadyEvent extends BaseEvent {
  type: 'agent.ready';
}

export interface AgentStateChangeEvent extends BaseEvent {
  type: 'agent.stateChange';
  state: AgentState;
  previousState: AgentState;
}

export interface AgentErrorEvent extends BaseEvent {
  type: 'agent.error';
  error: Error;
  recoverable: boolean;
  context?: string;
}

export type AgentEvent = AgentReadyEvent | AgentStateChangeEvent | AgentErrorEvent;

/**
 * Audio events
 */
export interface AudioCaptureStartEvent extends BaseEvent {
  type: 'audio.capture.start';
}

export interface AudioCaptureStopEvent extends BaseEvent {
  type: 'audio.capture.stop';
}

export interface AudioCaptureErrorEvent extends BaseEvent {
  type: 'audio.capture.error';
  error: Error;
}

export interface AudioPlaybackStartEvent extends BaseEvent {
  type: 'audio.playback.start';
}

export interface AudioPlaybackEndEvent extends BaseEvent {
  type: 'audio.playback.end';
}

export interface AudioPlaybackErrorEvent extends BaseEvent {
  type: 'audio.playback.error';
  error: Error;
}

export type AudioEvent =
  | AudioCaptureStartEvent
  | AudioCaptureStopEvent
  | AudioCaptureErrorEvent
  | AudioPlaybackStartEvent
  | AudioPlaybackEndEvent
  | AudioPlaybackErrorEvent;

/**
 * All possible events
 */
export type CompositeVoiceEvent =
  | TranscriptionEvent
  | LLMEvent
  | TTSEvent
  | AgentEvent
  | AudioEvent;

/**
 * Event type string union
 */
export type EventType = CompositeVoiceEvent['type'];

/**
 * Event listener function type
 */
export type EventListener<T extends CompositeVoiceEvent = CompositeVoiceEvent> = (
  event: T
) => void | Promise<void>;

/**
 * Event listener map for typed event subscriptions
 */
export interface EventListenerMap {
  'transcription.start': EventListener<TranscriptionStartEvent>;
  'transcription.interim': EventListener<TranscriptionInterimEvent>;
  'transcription.final': EventListener<TranscriptionFinalEvent>;
  'transcription.error': EventListener<TranscriptionErrorEvent>;

  'llm.start': EventListener<LLMStartEvent>;
  'llm.chunk': EventListener<LLMChunkEvent>;
  'llm.complete': EventListener<LLMCompleteEvent>;
  'llm.error': EventListener<LLMErrorEvent>;

  'tts.start': EventListener<TTSStartEvent>;
  'tts.audio': EventListener<TTSAudioEvent>;
  'tts.metadata': EventListener<TTSMetadataEvent>;
  'tts.complete': EventListener<TTSCompleteEvent>;
  'tts.error': EventListener<TTSErrorEvent>;

  'agent.ready': EventListener<AgentReadyEvent>;
  'agent.stateChange': EventListener<AgentStateChangeEvent>;
  'agent.error': EventListener<AgentErrorEvent>;

  'audio.capture.start': EventListener<AudioCaptureStartEvent>;
  'audio.capture.stop': EventListener<AudioCaptureStopEvent>;
  'audio.capture.error': EventListener<AudioCaptureErrorEvent>;
  'audio.playback.start': EventListener<AudioPlaybackStartEvent>;
  'audio.playback.end': EventListener<AudioPlaybackEndEvent>;
  'audio.playback.error': EventListener<AudioPlaybackErrorEvent>;
}
