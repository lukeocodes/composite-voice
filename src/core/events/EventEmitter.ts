/**
 * Type-safe event emitter for CompositeVoice SDK
 */

import type { EventType, EventListener, EventListenerMap, CompositeVoiceEvent } from './types';

/**
 * Type-safe event emitter with support for wildcard listeners
 */
export class EventEmitter {
  private listeners: Map<EventType | '*', Set<EventListener>>;
  private maxListeners: number;

  constructor(maxListeners = 100) {
    this.listeners = new Map();
    this.maxListeners = maxListeners;
  }

  /**
   * Register an event listener
   * @param event Event type to listen for, or '*' for all events
   * @param listener Listener function
   * @returns Unsubscribe function
   */
  on<T extends EventType>(
    event: T | '*',
    listener: T extends '*' ? EventListener : EventListenerMap[T]
  ): () => void {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }

    const eventListeners = this.listeners.get(event)!;

    if (eventListeners.size >= this.maxListeners) {
      console.warn(
        `Warning: Possible EventEmitter memory leak detected. ${eventListeners.size} listeners added for event "${event}". ` +
          `Use emitter.setMaxListeners() to increase limit.`
      );
    }

    eventListeners.add(listener as EventListener);

    // Return unsubscribe function
    return () => this.off(event, listener);
  }

  /**
   * Register a one-time event listener
   * @param event Event type to listen for
   * @param listener Listener function
   * @returns Unsubscribe function
   */
  once<T extends EventType>(event: T, listener: EventListenerMap[T]): () => void {
    const wrappedListener = ((evt: CompositeVoiceEvent) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.off(event, wrappedListener as any);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      void listener(evt as any);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    }) as any;

    return this.on(event, wrappedListener);
  }

  /**
   * Remove an event listener
   * @param event Event type
   * @param listener Listener function to remove
   */
  off<T extends EventType>(
    event: T | '*',
    listener: T extends '*' ? EventListener : EventListenerMap[T]
  ): void {
    const eventListeners = this.listeners.get(event);
    if (eventListeners) {
      eventListeners.delete(listener as EventListener);
      if (eventListeners.size === 0) {
        this.listeners.delete(event);
      }
    }
  }

  /**
   * Remove all listeners for an event, or all listeners if no event specified
   * @param event Optional event type to remove listeners for
   */
  removeAllListeners(event?: EventType | '*'): void {
    if (event) {
      this.listeners.delete(event);
    } else {
      this.listeners.clear();
    }
  }

  /**
   * Emit an event to all registered listeners
   * @param event Event object to emit
   */
  async emit<T extends CompositeVoiceEvent>(event: T): Promise<void> {
    const eventListeners = this.listeners.get(event.type);
    const wildcardListeners = this.listeners.get('*');

    const allListeners = [...(eventListeners || []), ...(wildcardListeners || [])];

    // Execute all listeners
    await Promise.all(
      allListeners.map(async (listener) => {
        try {
          await listener(event);
        } catch (error) {
          console.error(`Error in event listener for "${event.type}":`, error);
        }
      })
    );
  }

  /**
   * Emit an event synchronously (doesn't wait for async listeners)
   * @param event Event object to emit
   */
  emitSync<T extends CompositeVoiceEvent>(event: T): void {
    const eventListeners = this.listeners.get(event.type);
    const wildcardListeners = this.listeners.get('*');

    const allListeners = [...(eventListeners || []), ...(wildcardListeners || [])];

    // Execute all listeners without awaiting
    for (const listener of allListeners) {
      try {
        void listener(event);
      } catch (error) {
        console.error(`Error in event listener for "${event.type}":`, error);
      }
    }
  }

  /**
   * Get the number of listeners for an event
   * @param event Event type
   * @returns Number of listeners
   */
  listenerCount(event: EventType | '*'): number {
    const eventListeners = this.listeners.get(event);
    return eventListeners ? eventListeners.size : 0;
  }

  /**
   * Get all event types that have listeners
   * @returns Array of event types
   */
  eventNames(): (EventType | '*')[] {
    return Array.from(this.listeners.keys());
  }

  /**
   * Set the maximum number of listeners per event
   * @param n Maximum number of listeners
   */
  setMaxListeners(n: number): void {
    this.maxListeners = n;
  }

  /**
   * Get the maximum number of listeners per event
   * @returns Maximum number of listeners
   */
  getMaxListeners(): number {
    return this.maxListeners;
  }
}
