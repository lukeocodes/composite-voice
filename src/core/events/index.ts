/**
 * Events module export
 */

export * from './types';
export * from './EventEmitter';
