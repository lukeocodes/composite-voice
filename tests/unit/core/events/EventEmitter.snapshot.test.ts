/**
 * EventEmitter snapshot tests
 */

import { EventEmitter } from '../../../../src/core/events/EventEmitter';
import type { AgentReadyEvent, TranscriptionFinalEvent } from '../../../../src/core/events/types';

describe('EventEmitter Snapshots', () => {
  it('should match event structure snapshot', () => {
    const event: AgentReadyEvent = {
      type: 'agent.ready',
      timestamp: 1234567890,
    };

    expect(event).toMatchSnapshot();
  });

  it('should match transcription event snapshot', () => {
    const event: TranscriptionFinalEvent = {
      type: 'transcription.final',
      text: 'Hello world',
      confidence: 0.95,
      timestamp: 1234567890,
      metadata: {
        model: 'nova-2',
        language: 'en-US',
      },
    };

    expect(event).toMatchSnapshot();
  });

  it('should match emitter state snapshot', () => {
    const emitter = new EventEmitter(5);

    emitter.on('agent.ready', () => {});
    emitter.on('transcription.final', () => {});
    emitter.on('*', () => {});

    const state = {
      maxListeners: emitter.getMaxListeners(),
      eventNames: emitter.eventNames(),
      listenerCounts: {
        'agent.ready': emitter.listenerCount('agent.ready'),
        'transcription.final': emitter.listenerCount('transcription.final'),
        '*': emitter.listenerCount('*'),
      },
    };

    expect(state).toMatchSnapshot();
  });
});
