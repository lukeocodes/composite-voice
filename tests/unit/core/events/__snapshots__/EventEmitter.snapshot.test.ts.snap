// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventEmitter Snapshots should match emitter state snapshot 1`] = `
{
  "eventNames": [
    "agent.ready",
    "transcription.final",
    "*",
  ],
  "listenerCounts": {
    "*": 1,
    "agent.ready": 1,
    "transcription.final": 1,
  },
  "maxListeners": 5,
}
`;

exports[`EventEmitter Snapshots should match event structure snapshot 1`] = `
{
  "timestamp": 1234567890,
  "type": "agent.ready",
}
`;

exports[`EventEmitter Snapshots should match transcription event snapshot 1`] = `
{
  "confidence": 0.95,
  "metadata": {
    "language": "en-US",
    "model": "nova-2",
  },
  "text": "Hello world",
  "timestamp": 1234567890,
  "type": "transcription.final",
}
`;
